File: creational/typescript/abstract_factory_structured/concrete_factories.ts
import { GUIFactory, DatabaseFactory, Button, Checkbox, IDatabaseConnection, DatabaseCommand } from './interfaces';
import { WindowsButton, WindowsCheckbox, MacButton, MacCheckbox, MySQLConnection, MySQLCommand, PostgreSQLConnection, PostgreSQLCommand } from './concrete_products';

export class WindowsFactory implements GUIFactory {
  createButton(): Button { return new WindowsButton(); }
  createCheckbox(): Checkbox { return new WindowsCheckbox(); }
}

export class MacFactory implements GUIFactory {
  createButton(): Button { return new MacButton(); }
  createCheckbox(): Checkbox { return new MacCheckbox(); }
}

export class MySQLFactory implements DatabaseFactory {
  createConnection(): IDatabaseConnection { return new MySQLConnection(); }
  createCommand(cmd: string): DatabaseCommand { return new MySQLCommand(cmd); }
}

export class PostgreSQLFactory implements DatabaseFactory {
  createConnection(): IDatabaseConnection { return new PostgreSQLConnection(); }
  createCommand(cmd: string): DatabaseCommand { return new PostgreSQLCommand(cmd); }
}

File: creational/dart/abstract_factory_structured/concrete_factories.dart
import 'interfaces.dart';
import 'concrete_products.dart';

class WindowsFactory extends GUIFactory {
  @override
  Button createButton() => WindowsButton();
  @override
  Checkbox createCheckbox() => WindowsCheckbox();
}

class MacFactory extends GUIFactory {
  @override
  Button createButton() => MacButton();
  @override
  Checkbox createCheckbox() => MacCheckbox();
}

class MySQLFactory extends DatabaseFactory {
  @override
  IDatabaseConnection createConnection() => MySQLConnection();
  @override
  DatabaseCommand createCommand(String cmd) => MySQLCommand(cmd);
}

class PostgreSQLFactory extends DatabaseFactory {
  @override
  IDatabaseConnection createConnection() => PostgreSQLConnection();
  @override
  DatabaseCommand createCommand(String cmd) => PostgreSQLCommand(cmd);
}
